### DISCUSSION TITLE
How to change python version?

### DISCUSSION THREAD

**Discussion by FrockSaints** [NanoComp/meep#2224] [Q&A] [answered]
Sounds simple but I am unsure of how to proceed.
I've built from source on Ubuntu 18.04 with the help of the automated script (https://meep.readthedocs.io/en/latest/Build_From_Source/#building-from-source), but the python version is 3.6.9, which apparently is the default for this version of Ubuntu. I need python 3.7 or higher to use Tensorflow (and I'm also using Colour, but that dependency can be done away with), so I installed python 3.8.

However, since meep was built with python 3.6, there's conflict in both directions.
I figured changing "python3" to "python3.8" in the following line in "build-meep.sh" and building it in another folder would suffice...
`autogensh --with-mpi --with-openmp --enable-single PYTHON=python3`

But the result is that it simply didn't create the "/lib/python3.8/site-packages/meep/" directory. (I previously added 
`export PYTHONPATH=/home/frock/meep/lib/python3.6/site-packages/` to .profile and that allowed me to import meep in python 3.6)

If there's not an easy fix I'll probably run meep and tensorflow separately on their respective python versions and hope the numpy save format hasn't changed...
Thanks.

**Comment by stevengj** [comment]
You can certainly change the Python path with simply `PYTHON=somepython`, as long as `somepython` is in your path.

Assuming you did `make install`, then it should install the meep module into your installation prefix, whatever that is.  What did you set `--prefix` to?

If you do `make -n install` (the `-n` means that it prints out the installation commands but doesn't actually do anything), if you look carefully at the output you will be able to see where the Python modules got installed.

**Reply by stevengj** [reply]
In particular, if you look at `make -n install | grep basedir`, it should be something like:
```
	                    --basedir "/some/path/lib/python3.7/site-packages/meep/adjoint" $dlist; \
	                    --basedir "/some/path/lib/python3.7/site-packages/meep" $dlist; \
	                    --basedir "/some/path/lib/python3.7/site-packages/meep/mpb" $dlist; \
```
where `/some/path` is replaced by your installation prefix.

**Comment by stevengj** [comment]
Note also that you need to ensure that numpy is installed for the Python version you are using, otherwise Meep won't build the Python components.

If you do `grep WITH_PYTHON config.status`, it should show
```
S["WITH_PYTHON_FALSE"]="#"
S["WITH_PYTHON_TRUE"]=""
```
if Python is enabled.

**Reply by FrockSaints** [reply]
I used this automated script (with the aforementioned changes to "build-meep.sh" before building) :
```
mkdir -p /where/to/install/meep
cd /where/to/install/meep
wget https://raw.githubusercontent.com/NanoComp/meep/master/contrib/build-meep.sh
chmod +x build-meep.sh
./build-meep.sh
```
`make -n install | grep basedir` (run inside `~/meep/src/meep/`) shows nothing. Is that because I didn't set `--prefix` manually?

`grep WITH_PYTHON config.status` returns the opposite, as in, Python was not enabled, it seems.

Regarding having numpy installed, I hadn't realized it doesn't install a new numpy and was simply accessing the old one. 
I reinstalled the OS (and did that again a few more times throughout the process, just to make sure) and tried installing numpy using pip (pip3 in this case), with `python3.8 -m pip install numpy`. Strange, because it doesn't recognize pip3 in there but it does if I use it directly (`pip3 install numpy`). Regardless, it failed. Said it was missing Cython so I tried adding it with pip but again more issues with dependencies arose. I tried purging python3.6 (purge, remove and autoremove) from the system and installing python3.8, and following the recommendations in the pip docs (https://pip.pypa.io/en/stable/installation/): 
`python3.8 -m ensurepip --upgrade` fails, so I tried running get-pip.py, which failed with `ModuleNotFoundError: No module named 'distutils.cmd'`. 
So I ran `sudo apt-get install python3-distutils`, and realized it installed python 3.6 back, at which point I flipped the metaphorical table.

I don't mind using another WSL OS. I already have Ubuntu 20.04 installed due to another project, and I was previously using the conda serial meep package in Debian 9.5. But the automated script works with Ubuntu 16.04 and 18.04, so I tried the latest of the two. I don't have the expertise to get it working on another OS from source...

On top of those worries, I also realized that I can't just do `mpirun -np 6 python3 run.py`, because I'm doing a hard spheres simulation to determine the centers of core-shells before appending the corresponding `mp.Sphere` objects to the geometry, and that runs a lot slower than serial. I wonder if it's doing 6 simulations instead...

Thanks for the help so far and for reading my inexperienced attempts at solving this issue.

**Reply by stevengj** [reply]
>  `make -n install | grep basedir` (run inside`~/meep/src/meep/`) shows nothing.

Probably it didn't find a usable Python installation (probably because you didn't install numpy for that Python version).

**Reply by FrockSaints** [reply]
Ok so I managed to get python3.8 _with_ numpy (this did the trick - https://stackoverflow.com/questions/63207385/how-do-i-install-pip-for-python-3-8-on-ubuntu-without-changing-any-defaults), and used `PYTHON=python3.8` (python3.8 works in the terminal and opens Python 3.18.13 and I can import numpy and `numpy.version.version` is 1.23.2). (Again, added the `--enable-single flag`.)
Yet after building meep, I run into the same issues as before...
I don't know what I'm doing wrong, or if I'm doing what I'm supposed to but there's some unforeseen issue...
Appreciate if you could lend your expertise once again, now that I've reached another dead end.

**Reply by FrockSaints** [reply]
I believe the issue is that it's not just numpy that's necessary, but also scipy and libpython. I managed to install scipy with pip but not libpython, and since the repository for Ubuntu 18.04 defaults to python 3.6, I'm again stuck in dependency hell...

**Comment by Yaraslaut** [comment, marked as answer]
I've been working on building script recently for my environment, and am planning to create PR soon, but in mean time you can try to get it directly [here](https://github.com/Yaraslaut/meep/blob/build-meep_script/contrib/build-meep.sh) and it supports newer versions of ubuntu

**Reply by FrockSaints** [reply]
Thanks for sharing!
I tried it (with `--enable-single`) on 20.04 and got this, followed by a bunch of warnings: `/usr/include/hdf5/openmpi/H5public.h:60:13: fatal error: mpi.h: No such file or directory`.

I'll try `with_guile=false`, which I believe does the job of `--without-hdf5`...? I don't need hdf5 for now so hopefully that fixes it, and I'll update on the situation regardless, but it's worth looking into the hdf5 issue regardless, althought it's probably me who messed something up. 
Sorry for the repeated edits, by the way. Wanted to update on the situation as soon as possible but was also in the middle of something else.

**Reply by FrockSaints** [reply]
It failed again, this time with the following:
```
/usr/include/hdf5/openmpi/H5public.h:60:13: fatal error: mpi.h: No such file or directory
   60 | #   include <mpi.h>
```
Seems like it's still depending on hdf5 and thus, for some reason, crashing.
Where should I add `--without-hdf5` to? CFLAGS? CPPFLAGS? Or somewhere else?
Appreciate the support.

**Reply by Yaraslaut** [reply]
You can try few things, you can try to install mpi, for example `mpi-default-dev` package for 20.04.
When I was trying to compile without hdf5 I was getting compilation errors, but you can add configuration setting to build-meep.sh script via variables like `CONFIGURE_meep` and then run `CONFIGURE_meep="--without-hdf5" sh build-meep.sh `
One additional thing that you can try is to specify `--without-mpi` flag for the script `sh build-meep.sh --without-mpi`
Overall i tested script on ubuntu 22.04 and everything was alright, i will check then on 20.04

**Reply by FrockSaints** [reply]
Actually it's more convenient for me to build on 22.04, as I can reset it at any time, while having other things installed in 20.04.
I just happened to try 20.04 first because I already had it installed and then forgot 22.04 is a thing (very tired).
Tried it on 22.04 and had to replace `guile-2.0-dev` with `$libguile`, but still got this `fatal error: mpi.h: No such file or directory` afterwards.
Unlike for 20.04, this was on a fresh install.
I'll try `--without-hdf5` on both later, and update on the situation.

**Reply by Yaraslaut** [reply]
> Actually it's more convenient for me to build on 22.04, as I can reset it at any time, while having other things installed in 20.04. I just happened to try 20.04 first because I already had it installed and then forgot 22.04 is a thing (very tired). Tried it on 22.04 and had to replace `guile-2.0-dev` with `$libguile`, but still got this `fatal error: mpi.h: No such file or directory` afterwards. Unlike for 20.04, this was on a fresh install. I'll try `--without-hdf5` on both later, and update on the situation.

Thanks for the script check, I fixed errors there and now it builds on fresh system

**Reply by FrockSaints** [reply]
Thank you very much!
Was stuck trying to make it work without hdf5 before but can confirm that it's working in Ubuntu 22.04 with hdf5 now.
I also tried `--enable-single` and that's working well. The only thing left to figure out is why parallel execution in my case is as slow or slower than serial, but that's better left for another discussion.

**Reply by Yaraslaut** [reply]
> Thank you very much! Was stuck trying to make it work without hdf5 before but can confirm that it's working in Ubuntu 22.04 with hdf5 now. I also tried `--enable-single` and that's working well. The only thing left to figure out is why parallel execution in my case is as slow or slower than serial, but that's better left for another discussion.

Welcome 
I also noticed weird behaviour of mpi and I personally use only openmp suport
